# Cleaning Ebay Car Sales Dataset
The aim of this notebook is to produce a clean dataset that is ready for data analysis

In [230]:
import pandas as pd
import numpy as np

autos = pd.read_csv('autos.csv', encoding='Latin-1')
autos.info()
autos.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
dateCrawled            50000 non-null object
name                   50000 non-null object
seller                 50000 non-null object
offerType              50000 non-null object
price                  50000 non-null object
abtest                 50000 non-null object
vehicleType            44905 non-null object
yearOfRegistration     50000 non-null int64
gearbox                47320 non-null object
powerPS                50000 non-null int64
model                  47242 non-null object
odometer               50000 non-null object
monthOfRegistration    50000 non-null int64
fuelType               45518 non-null object
brand                  50000 non-null object
notRepairedDamage      40171 non-null object
dateCreated            50000 non-null object
nrOfPictures           50000 non-null int64
postalCode             50000 non-null int64
lastSeen               50000 non-null obj

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

## Clean up columns
| Original | New |
|----------|-----|
| yearOfRegistration | registration_year|
| monthOfRegistration | registation_month |
| notRepairedDamage | Unrepaired damage |
| dateCreated | ad_created |
And change the rest to camel case. This is done for clarity and to provide column names which follows Python convention.

In [231]:
print(autos.columns)

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')


In [232]:
autos.columns = ['date_crawled', 'name', 'seller', 'offer_type', 'price', 'ab_test',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'number_of_pictures', 'postal_code',
       'last_seen']
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price  ab_test vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_ps   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  number_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00                   0   
1         bmw              nein  2016-04-04 00:00:00                   0   
2  volkswagen              nein  2016-03-26 00:00:00                   0   
3       smart              nein  2016-03-12 00:00:00                   0   
4        ford              nein  2016-04-01 00:00:00                   0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

## Explore the dataset

In [233]:
autos.describe(include='all')

date_crawled         name  seller offer_type  price ab_test  \
count                 50000        50000   50000      50000  50000   50000   
unique                48213        38754       2          2   2357       2   
top     2016-03-09 11:54:38  Ford_Fiesta  privat    Angebot     $0    test   
freq                      3           78   49999      49999   1421   25756   
mean                    NaN          NaN     NaN        NaN    NaN     NaN   
std                     NaN          NaN     NaN        NaN    NaN     NaN   
min                     NaN          NaN     NaN        NaN    NaN     NaN   
25%                     NaN          NaN     NaN        NaN    NaN     NaN   
50%                     NaN          NaN     NaN        NaN    NaN     NaN   
75%                     NaN          NaN     NaN        NaN    NaN     NaN   
max                     NaN          NaN     NaN        NaN    NaN     NaN   

       vehicle_type  registration_year  gearbox      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  number_of_pictures   postal_code  \
count                 50000             50000.0  50000.000000   
unique                   76                 NaN           NaN   
top     2016-04-03 00:00:00                 NaN           NaN   
freq                   1946                 NaN           NaN   
mean                    NaN                 0.0  50813.627300   
std                     NaN                 0.0  25779.747957   
min                     NaN                 0.0   1067.000000   
25%                     NaN                 0.0  30451.000000   
50%                     NaN                 0.0  49577.000000   
75%                     NaN                 0.0  71540.000000   
max                     NaN                 0.0  99998.000000   

                  last_seen  
count                 50000  
unique                39481  
top     2016-04-07 06:17:27  
freq                      8  
mean                    NaN  
std                     NaN  
min                     NaN  
25%                     NaN  
50%                     NaN  
75%                     NaN  
max                     NaN

Columns with two or less unique values
- seller
- offer_type
- ab_test
- gearbox
- unrepaired_damage

Columns that are numeric but represented as string
- price
- odometer

Columns that need to be investigated more:
- registration_year
- registration_month

## Cleanup

In [234]:
autos['odometer'].unique()

array(['150,000km', '70,000km', '50,000km', '80,000km', '10,000km',
       '30,000km', '125,000km', '90,000km', '20,000km', '60,000km',
       '5,000km', '100,000km', '40,000km'], dtype=object)

In [235]:
autos['odometer'] = (autos['odometer']
                    .str.replace('km', '')
                    .str.replace(',',''))
autos['odometer'] = autos['odometer'].astype(int)
autos.rename({'odometer': 'odometer_km'}, axis=1, inplace=True)
autos.describe()

registration_year      power_ps    odometer_km  registration_month  \
count       50000.000000  50000.000000   50000.000000        50000.000000   
mean         2005.073280    116.355920  125732.700000            5.723360   
std           105.712813    209.216627   40042.211706            3.711984   
min          1000.000000      0.000000    5000.000000            0.000000   
25%          1999.000000     70.000000  125000.000000            3.000000   
50%          2003.000000    105.000000  150000.000000            6.000000   
75%          2008.000000    150.000000  150000.000000            9.000000   
max          9999.000000  17700.000000  150000.000000           12.000000   

       number_of_pictures   postal_code  
count             50000.0  50000.000000  
mean                  0.0  50813.627300  
std                   0.0  25779.747957  
min                   0.0   1067.000000  
25%                   0.0  30451.000000  
50%                   0.0  49577.000000  
75%                   0.0  71540.000000  
max                   0.0  99998.000000

In [236]:
autos['price'].unique()

array(['$5,000', '$8,500', '$8,990', ..., '$385', '$22,200', '$16,995'],
      dtype=object)

In [237]:
autos['price'] = (autos['price']
                 .str.replace('$', '')
                 .str.replace(',', ''))
autos['price'] = autos['price'].astype(float)
autos['price'].describe()

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64

In [238]:
autos['odometer_km'].unique()

array([150000,  70000,  50000,  80000,  10000,  30000, 125000,  90000,
        20000,  60000,   5000, 100000,  40000])

In [239]:
autos['odometer_km'].describe()

count     50000.000000
mean     125732.700000
std       40042.211706
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64

In [240]:
autos['odometer_km'].value_counts()

150000    32424
125000     5170
100000     2169
90000      1757
80000      1436
70000      1230
60000      1164
50000      1027
5000        967
40000       819
30000       789
20000       784
10000       264
Name: odometer_km, dtype: int64

In [241]:
autos['price'].unique()

array([ 5000.,  8500.,  8990., ...,   385., 22200., 16995.])

In [242]:
autos['price'].describe()

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64

In [243]:
autos['price'].unique().shape

(2357,)

In [244]:
autos['price'].value_counts().sort_index(ascending=True)

0.0           1421
1.0            156
2.0              3
3.0              1
5.0              2
8.0              1
9.0              1
10.0             7
11.0             2
12.0             3
13.0             2
14.0             1
15.0             2
17.0             3
18.0             1
20.0             4
25.0             5
29.0             1
30.0             7
35.0             1
40.0             6
45.0             4
47.0             1
49.0             4
50.0            49
55.0             2
59.0             1
60.0             9
65.0             5
66.0             1
              ... 
151990.0         1
155000.0         1
163500.0         1
163991.0         1
169000.0         1
169999.0         1
175000.0         1
180000.0         1
190000.0         1
194000.0         1
197000.0         1
198000.0         1
220000.0         1
250000.0         1
259000.0         1
265000.0         1
295000.0         1
299000.0         1
345000.0         1
350000.0         1
999990.0         1
999999.0    

Some of these prices are unrealistic. We have outliers. Arbitrarily, I'm selecting a value of $2,000 - $20,000 as valid prices.

In [245]:
autos = autos[autos['price'].between(1999,15000)]
autos.describe()

price  registration_year      power_ps    odometer_km  \
count  26088.000000       26088.000000  26088.000000   26088.000000   
mean    6009.711285        2005.863002    127.830037  124411.990187   
std     3434.960458          76.557763    237.013216   38458.395471   
min     1999.000000        1001.000000      0.000000    5000.000000   
25%     3150.000000        2002.000000     87.000000  100000.000000   
50%     4999.000000        2005.000000    120.000000  150000.000000   
75%     8000.000000        2008.000000    155.000000  150000.000000   
max    15000.000000        9999.000000  17700.000000  150000.000000   

       registration_month  number_of_pictures   postal_code  
count        26088.000000             26088.0  26088.000000  
mean             6.070147                 0.0  51362.610626  
std              3.556679                 0.0  25715.057852  
min              0.000000                 0.0   1067.000000  
25%              3.000000                 0.0  30938.000000  
50%              6.000000                 0.0  50735.000000  
75%              9.000000                 0.0  72202.000000  
max             12.000000                 0.0  99998.000000

By getting rid of the outliers, we also got rid of about half of the data

### Date Crawled, Ad Created, Last Seen distribution by percentage

In [246]:
date_crawled_distribution = autos['date_crawled'].str[:10].value_counts(normalize=True,dropna=False)*100
date_crawled_distribution.sort_index(ascending=False)
date_crawled_distribution.head()

2016-04-03    3.948175
2016-03-12    3.871512
2016-03-14    3.775682
2016-03-20    3.752683
2016-04-04    3.695186
Name: date_crawled, dtype: float64

The data is well balanced

In [247]:
ad_created_distribution = autos['ad_created'].str[:10].value_counts(normalize=True,dropna=False)*100
ad_created_distribution.sort_index(ascending=False)
ad_created_distribution.head()

2016-04-03    3.982674
2016-03-12    3.825514
2016-03-20    3.764183
2016-04-04    3.722018
2016-03-21    3.702852
Name: ad_created, dtype: float64

In [248]:
last_seen_distribution = autos['last_seen'].str[:10].value_counts(normalize=True,dropna=False)*100
last_seen_distribution.sort_index(ascending=False)
last_seen_distribution.head()

2016-04-06    23.593223
2016-04-07    14.198099
2016-04-05    13.101809
2016-03-17     2.637228
2016-04-03     2.495400
Name: last_seen, dtype: float64

Almost 25% of the ads has been seen last on April 7.

In [249]:
autos['registration_year'].describe()

count    26088.000000
mean      2005.863002
std         76.557763
min       1001.000000
25%       2002.000000
50%       2005.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

We have outliers on the `registration_year` column. We're not in the year 9999 yet and cars were not yet invented in 1001.

We can probably accept data from 1960 to 2016. It's a bit on the safe side, but maybe there are people who are selling their old cars? 2016 is the year when this dataset was created to that's the maximum value.

In [250]:
cars_1960_to_2016 = autos['registration_year'].between(1960, 2016)
autos = autos[cars_1960_to_2016]
autos.describe()

price  registration_year      power_ps    odometer_km  \
count  25157.000000       25157.000000  25157.000000   25157.000000   
mean    6037.144055        2004.460150    128.037286  124383.471797   
std     3440.837232           6.281307    218.146623   38385.015710   
min     1999.000000        1960.000000      0.000000    5000.000000   
25%     3199.000000        2002.000000     90.000000  100000.000000   
50%     4999.000000        2005.000000    120.000000  150000.000000   
75%     8150.000000        2008.000000    155.000000  150000.000000   
max    15000.000000        2016.000000  17700.000000  150000.000000   

       registration_month  number_of_pictures   postal_code  
count        25157.000000             25157.0  25157.000000  
mean             6.113209                 0.0  51518.710220  
std              3.541654                 0.0  25701.032915  
min              0.000000                 0.0   1067.000000  
25%              3.000000                 0.0  31134.000000  
50%              6.000000                 0.0  50827.000000  
75%              9.000000                 0.0  72336.000000  
max             12.000000                 0.0  99998.000000

In [251]:
autos['registration_year'].value_counts(normalize=True).sort_values(ascending=False).head(10)

2005    0.097150
2006    0.096911
2004    0.086417
2007    0.080773
2008    0.075128
2003    0.072465
2009    0.068609
2002    0.056525
2001    0.048416
2010    0.045912
Name: registration_year, dtype: float64

People seem to be selling cars they bought 10-ish years ago

## Aggregation

In [252]:
autos['brand'].unique()

array(['peugeot', 'bmw', 'volkswagen', 'smart', 'chrysler', 'renault',
       'mazda', 'audi', 'mini', 'mercedes_benz', 'ford', 'seat', 'opel',
       'dacia', 'jeep', 'volvo', 'nissan', 'toyota', 'skoda',
       'sonstige_autos', 'fiat', 'mitsubishi', 'porsche', 'kia',
       'citroen', 'suzuki', 'honda', 'hyundai', 'chevrolet', 'subaru',
       'land_rover', 'jaguar', 'alfa_romeo', 'daihatsu', 'saab',
       'trabant', 'daewoo', 'lancia', 'lada', 'rover'], dtype=object)

In [253]:
autos['brand'].value_counts(normalize=True)

volkswagen        0.215606
bmw               0.129387
mercedes_benz     0.109433
audi              0.094288
opel              0.085265
ford              0.055889
renault           0.029813
peugeot           0.028740
skoda             0.023135
fiat              0.021107
smart             0.018643
seat              0.017689
toyota            0.017172
mazda             0.015185
citroen           0.015026
nissan            0.013634
mini              0.012521
hyundai           0.012323
sonstige_autos    0.009739
volvo             0.008626
kia               0.008427
honda             0.007950
chevrolet         0.007632
suzuki            0.006599
alfa_romeo        0.006440
mitsubishi        0.006320
dacia             0.004293
chrysler          0.003339
jeep              0.002981
porsche           0.002584
land_rover        0.001948
jaguar            0.001590
saab              0.001511
subaru            0.001511
daihatsu          0.001232
trabant           0.000755
lada              0.000636
l

Let's choose the top 20 brands in the above list

In [254]:
selected_brands = ['volkswagen', 'bmw', 'mercedes_benz', 'audi', 'opel', 'ford', 'renault', 'peugeot',
                  'skoda', 'fiat', 'seat', 'smart', 'toyota', 'mazda', 'mini', 'citroen', 'nissan',
                  'hyundai', 'sonstige_autos', 'porsche']

In [255]:
selected_brands_mean_price = {}
selected_brands_mean_mileage = {}

for brand in selected_brands:
    mean = autos.loc[autos['brand'] == brand, 'price'].mean()
    mileage_mean = autos.loc[autos['brand'] == brand, 'odometer_km'].mean()
    selected_brands_mean_price[brand] = mean
    selected_brands_mean_mileage[brand] = mileage_mean
    
selected_brands_mean_mileage

{'audi': 139222.17537942666,
 'bmw': 138281.1059907834,
 'citroen': 107341.26984126984,
 'fiat': 97815.44256120527,
 'ford': 118413.94025604552,
 'hyundai': 98790.32258064517,
 'mazda': 118468.58638743455,
 'mercedes_benz': 137384.67126770795,
 'mini': 101206.3492063492,
 'nissan': 107959.1836734694,
 'opel': 119815.85081585082,
 'peugeot': 115691.56293222684,
 'porsche': 126846.15384615384,
 'renault': 111353.33333333333,
 'seat': 114977.52808988764,
 'skoda': 110635.73883161512,
 'smart': 89285.71428571429,
 'sonstige_autos': 93102.04081632652,
 'toyota': 112233.79629629629,
 'volkswagen': 128156.34218289086}

In [256]:
selected_brands_mean_price

{'audi': 7015.847807757167,
 'bmw': 6870.379723502304,
 'citroen': 5133.656084656084,
 'fiat': 4981.088512241055,
 'ford': 5308.256045519203,
 'hyundai': 6187.493548387097,
 'mazda': 5453.379581151832,
 'mercedes_benz': 6353.35742826008,
 'mini': 8132.361904761905,
 'nissan': 6425.99416909621,
 'opel': 4988.60979020979,
 'peugeot': 4681.449515905947,
 'porsche': 10530.892307692307,
 'renault': 4708.586666666667,
 'seat': 5806.76404494382,
 'skoda': 6466.04295532646,
 'smart': 4469.63539445629,
 'sonstige_autos': 7363.579591836735,
 'toyota': 5578.898148148148,
 'volkswagen': 6087.15081120944}

In [257]:
price_mean = pd.Series(selected_brands_mean_price)
mileage_mean = pd.Series(selected_brands_mean_mileage)
# df = pd.DataFrame(mileage_mean, columns=['mean_mileage'])
df = pd.DataFrame(price_mean, columns=['mean_price'])
df['mean_mileage'] = mileage_mean

df

mean_price   mean_mileage
audi             7015.847808  139222.175379
bmw              6870.379724  138281.105991
citroen          5133.656085  107341.269841
fiat             4981.088512   97815.442561
ford             5308.256046  118413.940256
hyundai          6187.493548   98790.322581
mazda            5453.379581  118468.586387
mercedes_benz    6353.357428  137384.671268
mini             8132.361905  101206.349206
nissan           6425.994169  107959.183673
opel             4988.609790  119815.850816
peugeot          4681.449516  115691.562932
porsche         10530.892308  126846.153846
renault          4708.586667  111353.333333
seat             5806.764045  114977.528090
skoda            6466.042955  110635.738832
smart            4469.635394   89285.714286
sonstige_autos   7363.579592   93102.040816
toyota           5578.898148  112233.796296
volkswagen       6087.150811  128156.342183

In [258]:
df.sort_values(by=['mean_price'], inplace=True, ascending=False)
df

mean_price   mean_mileage
porsche         10530.892308  126846.153846
mini             8132.361905  101206.349206
sonstige_autos   7363.579592   93102.040816
audi             7015.847808  139222.175379
bmw              6870.379724  138281.105991
skoda            6466.042955  110635.738832
nissan           6425.994169  107959.183673
mercedes_benz    6353.357428  137384.671268
hyundai          6187.493548   98790.322581
volkswagen       6087.150811  128156.342183
seat             5806.764045  114977.528090
toyota           5578.898148  112233.796296
mazda            5453.379581  118468.586387
ford             5308.256046  118413.940256
citroen          5133.656085  107341.269841
opel             4988.609790  119815.850816
fiat             4981.088512   97815.442561
renault          4708.586667  111353.333333
peugeot          4681.449516  115691.562932
smart            4469.635394   89285.714286

The data doesn't show a correlation between the mileage and the price, it seems.

Data cleaning next steps:
- Identify categorical data that uses german words, translate them and map the values to their english counterparts
- Convert the dates to be uniform numeric data, so "2016-03-21" becomes the integer 20160321.
See if there are particular keywords in the name column that you can extract as new columns

Analysis next steps:
- Find the most common brand/model combinations
- Split the odometer_km into groups, and use aggregation to see if average prices follows any patterns based on the milage.
- How much cheaper are cars with damage than their non-damaged counterparts?